### PR TITLE
Replace popup menu with new UI

### LINKED ISSUE #686: Implement popupMenu with improved UI
__What should we do?__
We want to replace the old popup UI menu with a new and improved UI, which includes searching. 

**To do:**
- [x] Integrate diagram-js-ui
- [x] Implement new UI and replace old popupMenu
    - Have backwards compatibility
    - Improve popup menu foundations
       - Allow grouping, descriptions, documentation url 
- [x] Prioritize and fix most pressing [known issues](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues)
    - Keyboard accessible
    - Popup menu container within canvas
    
 Initial implementation concluded by https://github.com/bpmn-io/diagram-js/pull/687
   
 **Identified bugs/improvements:**
- [x] https://github.com/bpmn-io/diagram-js/issues/695
   
__Why should we do it?__
Preparation to support append anything - see https://github.com/bpmn-io/bpmn-js/issues/1627

---
Child of https://github.com/bpmn-io/bpmn-js/issues/1627

### PR DESCRIPTION
**Pull request by smbea**
Incorporates a reworked popup menu UI based on the [create/append anything prototype](https://github.com/bpmn-io/bpmn-js-connectors-extension).

This can be tested with: 

```
npx @bpmn-io/sr bpmn-io/bpmn-js#develop -l bpmn-io/diagram-js#replace-popup-menu
```

Note: to see a draft of the e2e result you can use this [bpmn-js branch](https://github.com/bpmn-io/bpmn-js/tree/wip-test-popup-menu) 

### New features
* Menu title
* Search
* Entry options: description, documentation url, group title
* Configurable width

### Behaviour changes
* popup menu exists outside of the canvas

### Breaking changes

* Overall UI (will not affect standard popup menus but might affect UI of menus with CSS customisations on top)
* New class `djs-parent` added to canvas root (`.djs-container`) and popup menu root 
  

#### Replace
<div>
Screenshot 2022-11-14 at 10 57 42
Screenshot 2022-11-08 at 22 05 09
</div>

#### Align
<div>
Screenshot 2022-11-14 at 10 57 59
Screenshot 2022-11-14 at 11 05 35
</div>

----

Closes #686

### CODE REVIEW

**Comment by philippfromme** [comment]
What's the state of this pull request? Looking at the comments in https://github.com/bpmn-io/diagram-js/issues/686 I'm not sure which ones have been addressed.

**Comment by philippfromme** [comment]
Do we want the search bar to be present in the popup menu for alignment?

image

**Comment by philippfromme** [comment]
> There are some configurable options that can be set when popupMenu is opened (in bpmn-js) and being searchable is one of them.

Is there a bpmn-js branch showing the usage of these options end-to-end?

**Review by philippfromme — COMMENTED** [review]
I found two things:

1. Increasing the `min-width` of the replace menu would prevent the width jumping when using the search

brave_PsEkMyurX3

2. The align menu doesn't look right; is this intended?

brave_8gpCpErzl6

**Comment by philippfromme** [comment]
> Actually in mine, I don't event get the last 2 options for some reason:

I think the bottom ones are [distribute](https://github.com/bpmn-io/bpmn-js/blob/develop/lib/features/distribute-elements/DistributeElementsMenuProvider.js) while the ones at the top are align.

**Comment by nikku** [comment]
We should use [the sketched UI](https://bpmn-io.github.io/bpmn-js-connectors-extension/?aa=1) as a reference (with regards to paddings / margins) and I'm sure we'll be able to adjust for "no search", "no-header" in the same spirit.

Happy to review this if you ping me @smbea once you feel like this is ready for review / stabilized UI wise.

**Comment by nikku** [comment]
Thanks @smbea for the context. I'll give it a review today.

**Comment by nikku** [comment]
I've scheduled a session to finalize our diagram-js-ui integration, together with @barmac @smbea and @philippfromme.

File: assets/diagram-js.css [hunk 1]
```diff
@@ -508,86 +508,305 @@ marker.djs-dragger tspan {
...
+
+.djs-popup .header-entry.selected {
+  background-color: var(--hover-background-color);
+}
+
+.djs-popup .search {
+  margin: 10px 12px;
+}
+
+.djs-popup .title {
```
Comment (nikku) [inline assets/diagram-js.css:616]: Do we want to keep the existing class names (where possible?)

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 2]
```diff
@@ -1,19 +1,9 @@
-import {
-  bootstrapDiagram,
-  getDiagramJS,
-  inject
-} from 'test/TestHelper';
-
-import testImage from './resources/a.png';
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:7]: Any reason we removed these tests? Is the capability gone now?

File: assets/diagram-js.css [hunk 3]
```diff
@@ -508,86 +508,305 @@ marker.djs-dragger tspan {
 /**
  * popup styles
  */
-.djs-popup .entry {
+ .djs-popup {
+  position: fixed;
```
Comment (nikku) [inline assets/diagram-js.css:512]: Should check formatting here.

File: assets/diagram-js.css [hunk 4]
```diff
@@ -508,86 +508,305 @@ marker.djs-dragger tspan {
...
+  top: 0;
+  left: 0;
+  z-index: 200;
+  line-height: 1;
+  font-family: "IBM Plex Sans", sans-serif;
+  background: var(--popup-background-color);
+  border: solid 1px var(--popup-border-color);
+  border-radius: 2px;
+
+  --color-white: #fff;
```
Comment (nikku) [inline assets/diagram-js.css:524]: Should we re-use colors defined globally here or move these colors to the global scope?

I'd love to move variable definitions to the global scope, too:

https://github.com/bpmn-io/diagram-js/blob/develop/assets/diagram-js.css#L5

+ prefix popup styles there with `--popup-`

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 5]
```diff
@@ -251,69 +242,34 @@ describe('features/popup', function() {
...
-      // when
-      popupMenu.open({}, 'menu' ,{ x: 100, y: 100 });
+      popupMenu.open({}, 'menu', { x: 100, y: 100 });
 
       var container = popupMenu._current.container;
+      var popup = container.children[0];
 
       // then
-      expect(getComputedStyle(container).visibility).not.to.eql('hidden');
+      expect(domClasses(container).has('djs-popup-parent')).to.be.true;
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:258]: We could make these tests more robust to verify that `domQueryAll('djs-popup', container)` returns a single element.

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 6]
```diff
@@ -251,69 +242,34 @@ describe('features/popup', function() {
...
+      var domEntry = queryEntry(popupMenu, 'entry2');
 
       expect(domEntry.textContent).to.eql('foo');
     }));
 
 
-    it('should add action-id to entry', inject(function(popupMenu) {
-
-      // when
-      popupMenu.registerProvider('item-menu', {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:291]: Did we remove grouping? Or was it superceeded by something?

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 7]
```diff
@@ -348,92 +304,6 @@ describe('features/popup', function() {
     }));
 
 
-    describe('grouping', function() {
-
-      it('should group entries', inject(function(popupMenu) {
-
-        // given
-        popupMenu.registerProvider('group-menu', {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:356]: Did we remove grouping? Or was it superceeded by something?

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 8]
```diff
@@ -728,385 +601,42 @@ describe('features/popup', function() {
...
-      ));
-
-    });
-
-  });
-
-
-  describe('menu styling', function() {
-
-    it('should add standard class to entry', inject(function(popupMenu) {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:757]: We could consider to restore that capability.

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 9]
```diff
@@ -728,385 +601,42 @@ describe('features/popup', function() {
...
-      // then
-      expect(domQueryAll('.entry', popupBody).length).to.eql(2);
-      expect(domQueryAll('.entry', popupHeader).length).to.eql(1);
-    }));
-  });
-
-
-  describe('singleton handling', function() {
-
-    it('should update the popup menu, when it is opened again' , inject(
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:845]: Why did we remove this?

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 10]
```diff
@@ -728,385 +601,42 @@ describe('features/popup', function() {
...
+        expect(spy).to.not.have.been.called;
+      }));
 
+    });
   });
 
 
   describe('header', function() {
 
-    it('should throw an error, if the id of a header entry is not set', inject(
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:898]: Why did we remove this?

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 11]
```diff
@@ -1143,19 +673,31 @@ describe('features/popup', function() {
...
+
+      // then
+      expect(function() {
+        popupMenu.open({}, 'menu');
+      }).to.throw('the position argument is missing');
+    }
+    ));
+
+
     it('should open within bounds above', inject(function(popupMenu, canvas) {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:695]: We don't need the canvas anymore I believe.

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 12]
```diff
@@ -1168,20 +710,21 @@ describe('features/popup', function() {
 
 
     it('should open within bounds above (limited client rect height)', inject(
+
       function(popupMenu, canvas) {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:721]: We don't need the canvas anymore I believe?

**Review by nikku — CHANGES_REQUESTED** [review]
Looks like great progress overall.

I love that you also tested internals ("the components"). Overall focus for tests shall remain the `PopupMenuSpec` though.

Some remarks:

* Do we want to keep existing classes (where they continue to fit)? Happy to ditch, but we should be explicit about this in the PR description; this is going to be a potential breaking change, and we need to document it.
* Do we want to keep the ability to show images in header entries?
* Do we want to keep existing grouping features? If not we must document an appropriate replacement ("categories")?

**Comment by nikku** [comment]
Thanks for https://github.com/bpmn-io/diagram-js/pull/687#issuecomment-1311460885 :tada:.

Would be really good if we could keep existing tests where we keep compatibility.

File: assets/diagram-js.css (hunk 4 above)
Comment (nikku) [reply assets/diagram-js.css:524]: Got cha! Maybe we find a way to re-use these styles / variables regardless, i.e. introduce a `djs-parent` selector that provides them to different roots (popup-menu + djs-container)?

**Review by nikku — DISMISSED** [review]
I was not able to do a full review of the latest changes, but did already go ahead and rebase the commit on top of latest develop.

@smbea Could you go ahead and create a CHANGELOG entry for the upcoming changes + populate the PR description with new changes:

* A screenshot to show the new styling
* Documentation of breaking changes (introduction of `.djs-parent` class), new replace menu behavior, new capabilities (search, title, ...).

We'd need that to ship this change.

**Review by nikku — COMMENTED** [review]
I'm done, after all. Looks awesome. The `PopupMenuSpec` is a great sentiment of not breaking existing contracts (unnecessarily).

----

Added the following changes on top:

* reorganized contribution -> first shared `parent`, then popup menu
* only render search if actually needed
* test teardown (PopupMenuComponent)
* global ESC key binding (without search!)
* recycling of state + callbacks (https://github.com/bpmn-io/diagram-js/pull/687/commits/5e0c57b11797aecbf2a577fae686127f1c49eacf)


From https://github.com/bpmn-io/diagram-js/pull/687/commits/2fcdb7dd5e286e464da2b77b22881619c41893de on I simplified the tests / brought them back to the original form with the purpose to have a functional DIFF in place. That allowed me to find the following not yet documented breaking change:

* `.djs-popup-header .entry` is now `.djs-popup .header .header-entry` => We could decide to restore the original classes (does not hurt?)
  
It also helped me find that `should only render body if entries exist` spec is removed. Shall we add a replacement?

---

All test formatting fixes can be squashed together or dropped as you wish.

**Pull request merged**